src/Tools/Auditor/Auditor.dfy

```diff
@@ -0,0 +1,105 @@
+include "../../AST/Entities.dfy"
+include "../../AST/Names.dfy"
+include "../../AST/Syntax.dfy"
+include "../../AST/Translator.Entity.dfy"
+include "../../Interop/CSharpDafnyASTModel.dfy"
+include "../../Interop/CSharpDafnyInterop.dfy"
+include "../../Utils/Library.dfy"
+include "Report.dfy"
+
+module {:extern "Bootstrap.Tools.Auditor"} {:options "-functionSyntax:4"} Bootstrap.Tools.Auditor {
+
+import opened AST.Entities
+import opened AST.Names
+import opened AST.Predicates
+import opened AST.Syntax.Exprs
+import E = AST.Translator.Entity
+import opened AuditReport
+import opened Interop
+import opened Utils.Lib.Datatypes
+import opened Utils.Lib.Seq
+
+//// AST traversals ////
+
+function {:opaque} Any<T>(P: T ~> bool, ts: seq<T>) : (b: bool)
+  reads P.reads
+  requires forall t | t in ts :: P.requires(t)
+  ensures b == exists t | t in ts :: P(t)
+  ensures b == exists i | 0 <= i < |ts| :: P(ts[i])
+{
+  !Seq.All(x reads P.reads requires P.requires(x) => !P(x), ts)
+}
+
+predicate IsAssumeStatement(e: Expr)
+{
+  && e.Apply?
+  && e.aop.Eager?
+  && e.aop.eOp.Builtin?
+  && e.aop.eOp.builtin.Predicate?
+  && e.aop.eOp.builtin.predTy.Assume?
+}
+
+predicate ContainsAssumeStatement(e: Expr)
+  decreases e.Depth()
+{
+  || Deep.Any_Expr(e, (c:Expr) => IsAssumeStatement(c))
+  || (&& e.Unsupported?
+      && Any((c:Expr) requires c.Depth() < e.Depth() => ContainsAssumeStatement(c), e.children))
+}
+
+//// Tag extraction and processing ////
+
+function TagIf(cond: bool, t: Tag): set<Tag> {
+  if cond then {t} else {}
+}
+
+function GetTags(e: Entity): set<Tag> {
+  TagIf(exists a | a in e.ei.attrs :: a.name == Extern, HasExternAttribute) +
+  TagIf(exists a | a in e.ei.attrs :: a.name == Axiom, HasAxiomAttribute) +
+  TagIf(e.Type? && e.t.SubsetType?, IsSubsetType) +
+  TagIf(e.Type? && e.t.NewType? && e.t.nt.pred.Some?, IsSubsetType) +
+  TagIf(e.Type? && e.t.SubsetType? && e.t.st.witnessExpr.None?, HasNoWitness) +
+  TagIf(e.Type? && e.t.NewType? && e.t.nt.witnessExpr.None?, HasNoWitness) +
+  TagIf(e.Definition? && e.d.Callable? && e.d.ci.body.None?, HasNoBody) +
+  TagIf(e.Definition? && e.d.Callable? && e.d.ci.body.Some? &&
+        ContainsAssumeStatement(e.d.ci.body.value), HasAssumeInBody) +
+  TagIf(e.Definition? && e.d.Callable? && |e.d.ci.ens| > 0, HasEnsuresClause) +
+  // TagIf(e.Definition? && e.d.Callable? && |e.d.ci.req| > 0, HasRequiresClause) +
+  TagIf(e.Definition? && e.d.Callable?, IsCallable)
```

Comment:
I might have used a record with one boolean field per tag rather than a set, but this is good too